lib/puppet/type/iis_site.rb

```diff
@@ -0,0 +1,270 @@
+      '
+    validate do |value|
+      unless value.kind_of?(Integer)
+        fail("Invalid value '#{value}'. Should be a number")
+      end
+      if value < 1048576 or value > 4294967295
+        fail("Invalid value '#{value}'. Cannot be less than 1048576 or greater than 4294967295")
+      end
+    end
+  end
+
+  newproperty(:loglocaltimerollover, :boolean => true) do
+    desc 'Use the system\'s local time to determine for the log file name as
+       as well as when the log file is rolled over'
+    newvalue(:true)
+    newvalue(:false)
+
+    munge do |value|
+      resource.munge_boolean(value)
+    end
+  end
+
+  newproperty(:logflags, :array_matching => :all) do
+    desc 'Specifies what W3C fields are logged in the IIS log file. This is only
+      valid when :logformat is set to W3C. '
+    validate do |value|
+      # if value.nil? or value.empty?
+      #   raise ArgumentError, "A non-empty defaultpage must be specified."
+      # end
+      unless value.kind_of?(Array) || value.kind_of?(String)
+        fail("Invalid value '#{value}'. Should be a string")
+      end
+      unless [
+        'Date','Time','ClientIP','UserName','SiteName','ComputerName','ServerIP',
+        'Method','UriStem','UriQuery','HttpStatus','Win32Status','BytesSent',
+        'BytesRecv','TimeTaken','ServerPort','UserAgent','Cookie','Referer',
+        'ProtocolVersion','Host','HttpSubStatus'
+        ].include?(value)
+        fail("Invalid value '#{value}'. Valid values are Date, Time, ClientIP,
+             UserName, SiteName, ComputerName, ServerIP,
+             Method, UriStem, UriQuery, HttpStatus, Win32Status, BytesSent,
+             BytesRecv, TimeTaken, ServerPort, UserAgent, Cookie, Referer,
+             ProtocolVersion, Host, HttpSubStatus")
+      end
+    end
+  end
+
+  validate do
+    # TODO: These need validation
+    # logperiod and logtruncatesize
+    # logtruncatesize needs int range validation
+    # logflags need to support 4 values and no value
+    # enabledprotocols need to support 2 values and no value
+
+    # TODO: need check if logperiod is not MaxSize if logtruncatesize is set. if not
+    if self[:logperiod] && self[:logtruncatesize]
+      fail("Cannot specify logperiod and logtruncatesize at the same time")
+    end
+
+    # can only use logflags if logformat is W3C
+    if self[:logflags]
+      unless ['W3C','w3c'].include?(self[:logformat])
+        fail("Cannot specify logflags when logformat is not W3C")
+      end
+    end
+    
+    if self[:serviceautostartprovidername]
+      fail("Must specify serviceautostartprovidertype as well as serviceautostartprovidername") unless self[:serviceautostartprovidertype]
+    end
+    
+    if self[:serviceautostartprovidertype]
+      fail("Must specify serviceautostartprovidername as well as serviceautostartprovidertype") unless self[:serviceautostartprovidername]
+    end
+
+    provider.validate if provider.respond_to?(:validate)
+  end
+
+  def munge_boolean(value)
```

Comment:
I once attempted this and ran into issues - and determined it only works for types built into Puppet for some reason. It could be I didn't have something set up properly - that was two years ago, so it's likely the require was missing.